### PR TITLE
Probabilities Module: User-defined policy for Boost Math instead of manual domain checks

### PR DESCRIPTION
**Pull request by fschopp**
_Please feel free to discuss and share your comments. However, please do not pull yet. I will consolidate Kevin's earlier pull request 131 and this one._

C++ AL:
- Boost global functions (so far, only error handlers) are now in dbal/BoostIntegration directory

Probability functions:
- Instead of isnan/isfinite/etc. checks, we now have a user-defined policy that propagates NaNs but errors out on all other boost domain errors (using the error message provided by boost).
- Boost probability functions are defined in modules/prob/boost.{hpp|cpp} using macros. New functions can now be added by adding only one line in both files.
- Provided a facility to override boost's domain checks (necessary, e.g., for the Fisher F-distribution).
- Student t-distribution now provides the standard boost interface.

Other modules:
- Adapted to the above changes

### CODE REVIEW

**Comment by cwelton** [comment]
Can you enumerate the boost issues and what is currently broken using the policy based implementation?

**Comment by fschopp** [comment]
All boost issues are aggregated in this file: https://github.com/fschopp/madlib/blob/prob_integrate/src/modules/prob/boost.hpp
Whenever I wrote a
`struct DomainCheck<DISITRUBTION>` specialization or a `DOMAIN_CHECK_OVERRIDE` macro, boost either has a bug or a behavior that we do not want (e.g., complaining about negative values passed to a CDF). I tried to document everything in the source code, and I linked to boost bugs I filed. Let me know if you have specific questions.

File: src/modules/prob/boost.cpp [hunk 1]
```diff
@@ -21,92 +16,120 @@
...
+DEFINE_CONTINUOUS_PROB_DISTR_2(beta, double, double)
+DEFINE_CONTINUOUS_PROB_DISTR_2(cauchy, double, double)
+DEFINE_CONTINUOUS_PROB_DISTR_1(chi_squared, double)
+DEFINE_CONTINUOUS_PROB_DISTR_1(exponential, double)
+DEFINE_CONTINUOUS_PROB_DISTR_2(extreme_value, double, double)
+DEFINE_CONTINUOUS_PROB_DISTR_2(fisher_f, double, double)
+DEFINE_CONTINUOUS_PROB_DISTR_2(gamma, double, double)
+DEFINE_CONTINUOUS_PROB_DISTR_2(inverse_chi_squared, double, double)
+DEFINE_CONTINUOUS_PROB_DISTR_2(inverse_gamma, double, double)
+// FIXME: Pending Boost bug 6934, we currently do not support the
```
Comment (cwelton) [inline src/modules/prob/boost.cpp]: `// FIXME: MADLIB-513: Pending Boost bug 6934 ...`

File: src/modules/prob/boost.hpp [hunk 2]
```diff
@@ -4,38 +4,530 @@
...
+    MADLIB_ITEM(non_central_chi_squared) \
+    MADLIB_ITEM(non_central_f) \
+    MADLIB_ITEM(non_central_t) \
+    MADLIB_ITEM(normal) \
+    MADLIB_ITEM(pareto) \
+    MADLIB_ITEM(rayleigh) \
+    MADLIB_ITEM(triangular) \
+    MADLIB_ITEM(uniform) \
+    MADLIB_ITEM(weibull)
+// FIXME: Pending Boost bug 6934, we currently do not support the
```
Comment (cwelton) [inline src/modules/prob/boost.hpp]: Again, pointer to MADLIB issue also desirable.

File: src/modules/prob/boost.hpp [hunk 3]
```diff
@@ -4,38 +4,530 @@
...
+#include <boost/math/distributions.hpp>
+
+namespace madlib {
+
+namespace modules {
+
+namespace prob {
+
+namespace {
+// No need to make this visiable beyond this translation unit.
```
Comment (cwelton) [inline src/modules/prob/boost.hpp]: s/visiable/visable/

File: src/modules/prob/boost.hpp [hunk 4]
```diff
@@ -4,38 +4,530 @@
...
+
 /**
- * @brief Chi-squared cumulative distribution function
+ * @brief Via (partial) spezialization, this class offers a way to override
```
Comment (cwelton) [inline src/modules/prob/boost.hpp]: s/spezialization/specialization/

**Comment by cwelton** [comment]
Overall looks good.  

Design looks reasonable.  A couple small comments and typos, but I don't see any glaring issues.  

The boost workarounds are a bit gross, but hopefully we can remove that after some upstream fixes.

File: src/modules/prob/boost.hpp (hunk 2 above)
Comment (fschopp) [inline src/modules/prob/boost.hpp]: Added.

File: src/modules/prob/boost.cpp (hunk 1 above)
Comment (fschopp) [inline src/modules/prob/boost.cpp]: Added.

**Pull request merged**